Base integration for thermostat preset modes

AlePerla opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Integrates (in a simple way, it could be further worked on to dynamically fetch both mode capabilities and available preset modes) the preset modes for thermostats

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

AlePerla commented:
Hi @midstar . I have worked a little as well on the integration to also add control for the preset mode. It is pretty much hardcoded, for sure it could be finer to fetch available modes dynamically from the cloud, yet I think it is a useful functionality. Feel free to suggest or make changes (unfortunately, I don't have much time to work on it)

midstar commented:
Hi. How come you have created the PR in "my clone" instead of the original HA core repo?

AlePerla commented:
I don't think it is ready for the original integration for the reasons I said above, I'm either not that experienced with the tuya SDK.
I don't know if you feel like working on it and then bring it to the original repo.
Otherwise, feel free to reject the PR, I'll keep it on mine until I'll be able to integrate the missing pieces

midstar commented:
Hi. Sorry for the late reply. The intention with this clone of the HA core repository is to include device specific parts in the code (currently defined in the file device_specific.py). In its latest design it is, according to myself, well prepared to be merged into the original repository. I will try to keep this clone up to date with the original repository and would prefer to not include other "experimental" parts in the code. However, I will keep this pull request open for other people to use your changes if they wish.

AlePerla commented:
You got a point.
If I'll have the chance to make it ready to be merged into the official repo, I'll take care of closing this PR!

epenet committed:
Adjust type hints for SirenEntityFeature (#82266)

epenet committed:
Adjust type hints for UpdateEntityFeature (#82268)

epenet committed:
Adjust type hints for VacuumEntityFeature (#82269)

epenet committed:
Adjust type hints for WaterHeaterEntityFeature (#82270)

bdraco committed:
Add websocket endpoints to control integration logging (#65158)

Co-authored-by: Erik Montnemery <erik@montnemery.com>
Co-authored-by: Erik <erik@montnemery.com>

bdraco committed:
Add bluetooth loggers to manifest.json (#82276)

bdraco committed:
Disable BLE options flow for sleepy shelly devices (#82283)

Noltari committed:
Update aioairzone to v0.5.0 (#82278)

Noltari committed:
Add QNAP QSW accumulated port sensors (#76514)

Co-authored-by: J. Nick Koston <nick@koston.org>

epenet committed:
Cleanup access to supported features (#82274)

bdraco committed:
Fix high latency from thundering heard at 0 microseconds (#82233)

* Fix high latency at 0 microseconds

fixes #82231

* fix async_track_utc_time_change alignment

* use replace to preserve fold

* naming

* tweak

* make async_fire_time_changed aware of the thundering heard issue

bdraco committed:
Offload platform specific bluetooth code to bluetooth-adapters (#82196)

* Offload platform specific bluetooth code to bluetooth-adapters

* adjust

* fix some more patch targets

* more test fixes

* almost there

* may not be setup yet

* more fixes

* fixes

* fix test

* fix merge

Aohzan committed:
Add intent_script debug logging (#82060)

jbouwh committed:
Add support for Mqtt protocol version 5 (#82260)

frenck committed:
Add number device class support to ESPHome (#82277)

cdce8p committed:
Cleanup pip cache workaround [ci] (#82272)

emontnemery committed:
Add config_entries.async_wait_component (#76980)

Co-authored-by: thecode <levyshay1@gmail.com>

daanbeverdam committed:
Fix moving average for 0 values (#80476)

daanbeverdam committed:
Fix moving average for 0 values (#80476)

jbouwh committed:
Remove legacy YAML support from MQTT (#82102)

bdraco committed:
Bump dbus-fast to 1.75.0 (#82289)

Commit:
[ci skip] Translation update

janiversen committed:
Solve Modbus reload issue (#82253)

fixes undefined

janiversen committed:
Solve Modbus reload issue (#82253)

fixes undefined

elupus committed:
Add kilo watts unit mapping for nibe_heatpump (#82284)

elupus committed:
Add kilo watts unit mapping for nibe_heatpump (#82284)

epenet committed:
Remove unused variable from homekit_controller (#82307)

epenet committed:
Cleanup supported_features default values (#82306)

yozik04 committed:
Add Vallox bypass locking switch entity (#75857)

* Switch entity

* adding missing function for tests

* Apply suggestions from code review

Co-authored-by: Sebastian Lövdahl <slovdahl@hibox.fi>

* Review

* fix

* Update homeassistant/components/vallox/switch.py

Co-authored-by: Andre Richter <andre-richter@users.noreply.github.com>

* Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

* Adding a separate parameter for the expected state

Co-authored-by: Sebastian Lövdahl <slovdahl@hibox.fi>
Co-authored-by: Andre Richter <andre-richter@users.noreply.github.com>
Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

jbouwh committed:
Improve type hints MQTT light schema json (#82208)

* Improve type hints schema json

* Add hint for brightness

* Follow up comments

* Follow up missed comments

* Correct hint on flash arg

* hints on one line

epenet committed:
Add precipitation device class (#81145)

bdraco committed:
Fix modbus test assuming repeat listeners always fire at 0 microseconds (#82320)

bdraco committed:
Fix litejet tests (#82324)

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

gjohansson-ST committed:
Code quality tests Wake on Lan (#82048)

gjohansson-ST committed:
Add unique id to Times of Day (#81196)

jbouwh committed:
Revert #81771 (wait_for_states) and #81801 (late review) (#82085)

* Revert "Address late review of config entry wait for states tests (#81801)"

This reverts commit 12d76a8a4f1cbc9403798a32dd335d928dd61373.

* Revert "Implement ConfigEntry async_wait_for_states (#81771)"

This reverts commit 3cc9ecf1dc0dc5fdace19770c076eb47d2f1c746.

frenck committed:
Remove CI cache cleanup (#82330)

hesselonline committed:
Add wallbox Energy Price (#82039)

fixes undefined

aykborstelmann committed:
AVM Fritz!Box SmartHome: Integrate Templates (#81885)

Noltari committed:
Update aioairzone to v0.5.1 (#82311)

Co-authored-by: J. Nick Koston <nick@koston.org>

Commit:
[ci skip] Translation update

bachya committed:
Bump `regenmaschine` to 2022.11.0 (#82337)

fixes undefined

bachya committed:
Bump `regenmaschine` to 2022.11.0 (#82337)

fixes undefined

bdraco committed:
Improve time change helper comments (#82349)

* Improve time change helper comments

https://github.com/home-assistant/core/pull/82324#discussion_r1026962806

* Improve time change helper comments

https://github.com/home-assistant/core/pull/82324#discussion_r1026962806

bbr111 committed:
Add Homematic sensor descriptions (#82156)

Update sensor.py

Add Sensor Descriptions.

Ernst79 committed:
Bump bthome-ble to 2.3.1 (#82342)

* Bump bthome-ble

* Bump bthome-ble

bdraco committed:
Bump flux_led to 0.28.34 (#82347)

* Bump flux_led to 0.28.33

fixes #75832

changelog: https://github.com/Danielhiversen/flux_led/compare/0.28.32...0.28.33

* more more bump for legacy turn on

bdraco committed:
Bump flux_led to 0.28.34 (#82347)

* Bump flux_led to 0.28.33

fixes #75832

changelog: https://github.com/Danielhiversen/flux_led/compare/0.28.32...0.28.33

* more more bump for legacy turn on

mib1185 committed:
Fix invalid configuration_url in Netatmo (#82372)

mib1185 committed:
Fix invalid configuration_url in Netatmo (#82372)

mib1185 committed:
Address late review of fritzbox (#82369)

rklomp committed:
Bump pysma to version 0.7.3 (#82343)

fixes undefined

rklomp committed:
Bump pysma to version 0.7.3 (#82343)

fixes undefined

elupus committed:
Use unit enums in nibe heatpump (#82370)

Use unit enums in nibe

thecode committed:
Add tests for Shelly binary_sensor platform (#82367)

* Add tests for Shelly binary_sensor platform

* Rename create to register for

flz committed:
iaqualink: fix supported_effects reference after function was renamed (#82383)

Fix supported_effects reference

thecode committed:
Add tests for Shelly sensor platform (#82385)

akx committed:
Add support for Sensirion BLE sensors (#82382)

bdraco committed:
Bump bleak-retry-connector to 2.8.5 (#82387)

changelog: https://github.com/Bluetooth-Devices/bleak-retry-connector/compare/v2.8.4...v2.8.5

from: https://github.com/esphome/esphome/pull/4049

bdraco committed:
Bump bleak-retry-connector to 2.8.5 (#82387)

changelog: https://github.com/Bluetooth-Devices/bleak-retry-connector/compare/v2.8.4...v2.8.5

from: https://github.com/esphome/esphome/pull/4049

marvin-w committed:
Update xknx to 1.2.1 (#82404)

marvin-w committed:
Update xknx to 1.2.1 (#82404)

TheJulianJES committed:
Bump PyViCare to 2.19.0 (#82381)

fixes undefined

TheJulianJES committed:
Bump PyViCare to 2.19.0 (#82381)

fixes undefined

dmulcahey committed:
Remove duplicate device class on Flo water sensor (#82407)

Fixes https://github.com/home-assistant/core/issues/82390
fixes undefined

PaarthShah committed:
Bump plexapi version to 4.13.1 (#82396)

akx committed:
Correct SensorDeviceClass for sensirion_ble (#82413)

bdraco committed:
Split bluetooth models into base_scanner and wrappers (#82291)

jbouwh committed:
Wait for MQTT entry on snips integration startup (#81733)

Wait for MQTT integration to load

bdraco committed:
Prevent powerwall from switching addresses if its online (#82410)

* Prevent powerwall from switching addresses if its online

If the wifi interface was discovered we would switch the
ip address in the entry to the wifi ip even if it was
connected via ethernet

* cover

* more cover

bdraco committed:
Prevent powerwall from switching addresses if its online (#82410)

* Prevent powerwall from switching addresses if its online

If the wifi interface was discovered we would switch the
ip address in the entry to the wifi ip even if it was
connected via ethernet

* cover

* more cover

dmulcahey committed:
Attempt to fix occasional Flo timeouts (#82408)

* Attempt to fix occasional Flo timeouts

* remove gather

dmulcahey committed:
Attempt to fix occasional Flo timeouts (#82408)

* Attempt to fix occasional Flo timeouts

* remove gather

akx committed:
Use HA SensorDeviceClass in ruuvitag-ble (#82321)

* Use HA SensorDeviceClass in ruuvitag-ble

Refs https://github.com/home-assistant/core/pull/81327#discussion_r1024565439

* Use HA SensorDeviceClass even more

bdraco committed:
Break out bluetooth apis into api.py (#82416)

* Break out bluetooth apis into api.py

Like #82291 this is not a functional change.

* cleanups

balloob committed:
Bumped version to 2022.11.4

felipediel committed:
Bump broadlink to 0.18.3 (#82427)

fixes undefined

cdce8p committed:
Fix round typing [shelly] (#82436)

cdce8p committed:
Fix round typing [accuweather] (#82433)

thecode committed:
Add tests coverage for Shelly entity (#82432)

* Add tests coverage for Shelly entity

* Make it generator expression

Commit:
[ci skip] Translation update

cdce8p committed:
Fix round typing [zwave_js] (#82439)

balloob committed:
2022.11.4 (#82428)

cdce8p committed:
Allow `None` return for notify `get_service` (#82444)

mvn23 committed:
Bump pyotgw to 2.1.3 (#82430)

cdce8p committed:
Fix round typing [tradfri] (#82438)

epenet committed:
Fix overkiz bitwise operation (#82456)

cdce8p committed:
Update pylint to 2.15.6 (#82440)

* Update pylint to 2.15.6

* Use single pylint disable

* Use implicit dict constructor

elupus committed:
Bump nibe to 1.3.0 (#82445)

cdce8p committed:
Code quality improvements for `nissan_leaf` (#82442)

Code quality improvements for nissan_leaf

jbouwh committed:
Improve type hints MQTT light schema template (#82211)

* Improve type hints MQTT light schema template

* A few improvements

* Follow up comments

cdce8p committed:
Code quality improvements for `dsmr` (#82443)

cdce8p committed:
Fix round typing [fritzbox] (#82434)

emontnemery committed:
Bump pychromecast to 13.0.1 (#82472)

epenet committed:
Fix suggested values in openuv config flow (#82479)

* Fix suggested values in openuv config flow

* Don't use new helper

thecode committed:
Add tests coverage for Shelly number platform (#82480)

cdce8p committed:
Fix round typing [isy994] (#82435)

cdce8p committed:
Fix round typing [tplink] (#82437)

cdce8p committed:
Fix picnic typing (#82476)

gjohansson-ST committed:
Add config flow to Scrape (#81193)

* Scrape take 2

* cleanup

* new entity name

* Fix name, add tests

* Use FlowResultType

* Add test abort

* hassfest

* Remove not needed test

* clean

* Remove config entry and implement datacoordinator

* fix codeowners

* fix codeowners

* codeowners reset

* Fix coordinator

* Remove test config_flow

* Fix tests

* hassfest

* reset config flow

* reset strings

* reset sensor

* Reconfig

* Fix tests

* coverage

* Remove coverage

* Remove print

* Add config flow

* Fix config flow

* Add back init

* Add entry to sensor

* float to int

* Fix SelectSelector

* Add tests for config entry to existing

* Test config flow

* Fix test reload

* Fix rebase

* Fix strings

* clean init

* Clean test_sensor

* Align sensor setup entry

* Add error to strings

* review changes

* clean tests

* Add back options flow

* review changes

* update_listener

* Add tests

* Remove duplicate abort

* strings

* sensors to sensor

* review changes

* more review changes

* clarify test payload

* fixture name

epenet committed:
Enforce AlarmControlPanelEntityFeature (#82313)

epenet committed:
Fix incorrect type hints in scrape (#82502)

dmulcahey committed:
Add config entities for lumi.motion.agl04 (#82087)

dmulcahey committed:
Add Aqara c1 pet feeder support to ZHA (#82340)

* Add Aqara c1 pet feeder support to ZHA

* clean up

* cleanup

* state classes for daily measurements

* cleanups

* cleanups

* restore the refreshing of the inverted value cache

* cleanup

wildekek committed:
Add new Amazon Polly voices (#82501)

GitHub-Action committed:
[ci skip] Translation update

puddly committed:
Bump ZHA dependencies (#82509)

* Bump ZHA dependencies

* Use the corrected `TypeValue` keyword argument name in unit tests

bdraco committed:
Accept advertisements from alternate scanners when a scanner stops scanning (#82448)

bdraco committed:
Show HomeKit Controller unhandled pairing error reason in the UI (#82505)

epenet committed:
Enforce CameraEntityFeature (#82325)

epenet committed:
Enforce ClimateEntityFeature (#82329)

epenet committed:
Enforce CoverEntityFeature (#82457)

* Enforce CoverEntityFeature

* Adjust pylint

epenet committed:
Enforce FanEntityFeature (#82458)

* Enforce FanEntityFeature

* Adjust pylint

epenet committed:
Enforce HumidifierEntityFeature (#82459)

epenet committed:
Enforce LightEntityFeature (#82460)

epenet committed:
Enforce LockEntityFeature (#82461)

epenet committed:
Enforce MediaPlayerEntityFeature (#82462)

epenet committed:
Enforce RemoteEntityFeature (#82463)

epenet committed:
Enforce SirenEntityFeature (#82464)

epenet committed:
Enforce UpdateEntityFeature (#82465)

epenet committed:
Enforce WaterHeaterEntityFeature (#82467)

epenet committed:
Enforce VacuumEntityFeature (#82466)

akx committed:
Configure autoflake for pre-commit (#82414)

Shutgun committed:
Migrate devolo Home Control to new entity naming (#74969)

janiversen committed:
Add full test of wrap option (#82308)

fixes undefined

emontnemery committed:
Migrate ZHA when enabling multi-PAN support on HA Yellow (#82213)

* Migrate ZHA when enabling multi-PAN support on HA Yellow

* Refactor BaseZhaFlow.async_step_maybe_confirm_ezsp_restore

* Change data passed to ZHA to initiate migration

* Catch errors during ZHA migration

* Fix ZhaMigrationHelper.async_prepare_yellow_migration return value

* Improve test coverage

* Improve test coverage

* Fix spelling

* Rename some none HA yellow specifics

* Rename again

* Increase number of migration retries + refactor

* Suppress OperationNotAllowed when reloading

* Adjust tests

emontnemery committed:
Sort ignore-words-list (#82520)

epenet committed:
Move PluggableAction to trigger helpers (#81900)

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>
Co-authored-by: Joakim Plate <elupus@ecce.se>
Co-authored-by: Shay Levy <levyshay1@gmail.com>

epenet committed:
Fix incorrect type hint in SchemaOptionsFlow (#82516)

yozik04 committed:
Add Vallox temperature control entities (#75858)

Co-authored-by: Sebastian Lövdahl <slovdahl@hibox.fi>
Co-authored-by: Andre Richter <andre-richter@users.noreply.github.com>
Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

jbouwh committed:
Add unique id option for `mqtt_room` sensors (#82521)

Add unique_id setting for mqtt_room

thecode committed:
Add tests coverage for Shelly climate platform (#82529)

Noltari committed:
Abort before trying to connect in QNAP QSW DHCP discovery when already configured (#82362)

epenet committed:
Improve type hints in acmedia config flow (#82534)

jesserockz committed:
Bump aioesphomeapi to 12.0.0 (#82543)

Jc2k committed:
Fix HomeKit CoAP connection getting RST incorrectly (#82553)

GitHub-Action committed:
[ci skip] Translation update

bdraco committed:
Ensure esphome client tasks await cancelation (#82547)

hesselonline committed:
Fix Wallbox price unit (#82420)

* Change Symbol to Currency code in test

* change unit to currency/kWh

* f string

ludeeus committed:
Add helper to access RepairsFlowManager (#82567)

async_external_step

frenck committed:
Not to Tox (#76582)

jbouwh committed:
Add Mqtt WebSockets  support (#82078)

* Add Mqtt WebSockets support

* Fix tests

* Add testing websockets options

* Add tests transport settings

* Do not use templates for ws_headers

* Use json helper - small corrections

elupus committed:
Bump to Arcam 1.0.1 and make strictly typed (#82487)

* Make arcam_fmj strictly typed

* Add test for invalid UDN

jbouwh committed:
Move mqtt device_tracker files (#82581)

Move mqtt device_tracker files to single file

matrixd2 committed:
Add yolink Dimmer support (#81970)

* Add yolink Dimmer support

* suggest change

* fix suggest

* fix suggest

* fix suggest

epenet committed:
Add type hints to template states (#82582)

* Add type hints to template states

* Undo rename

* Remove invalid mypy issue link

DCSBL committed:
Add support for controlling HomeWizard Energy Socket status light level (#82378)

* Add control option for Energy Socket status LED

* Add tests for number

* Fix failing tests in switch

* Fix comments

* Adjust name

* Simplify device_info

frenck committed:
Cleanup Climacell translations (#82584)

akx committed:
Make hassfest strictly typed (#82091)

gjohansson-ST committed:
Fully type min_max (#79496)

vigonotion committed:
Add yamaha brand configuration (#81267)

Co-authored-by: Franck Nijhof <git@frenck.dev>

jbouwh committed:
Add type hints on `Template().__init__()` (#82574)

epenet committed:
Add type hints to template tests (#82566)

epenet committed:
Fix TemplateError definition (#82570)

Noltari committed:
Add Airzone DHCP discovery support (#82339)

elupus committed:
Upgrade arcam_fmj to entity naming (#82586)

elupus committed:
Test corrections noticed during pytest upgrade (#82579)

* Withing trigger a call to coordinator after init

* Stop discovery task on STOP event

* Stop dsmr connection task on STOP

* Use autospec in modem_serial tests

close on AioSerial is non async, and should not be mocked as a coroutine

* Make sure responder is specced correctly

tests/components/emulated_hue/test_init.py::test_setup_works
  /Users/joakim/src/hass/home-assistant/homeassistant/components/emulated_hue/__init__.py:119: RuntimeWarning: coroutine 'AsyncMockMixin._execute_mock_call' was never awaited
    protocol.close()

* Don't assume Pymodbus is the only thing logging

* Use gather instead of wait

```
homeassistant/components/homeassistant/__init__.py:208: DeprecationWarning: The explicit passing of coroutine objects to asyncio.wait() is deprecated since Python 3.8, and scheduled for removal in Python 3.11.
```

* Increase wait time for automation tests

Assuming exit within 1 seconds is unreasonable on a potentially loaded
test machine.

* Avoid changing app state after startup

eifinger committed:
Use assignment expression for alexa init (#81242)

epenet committed:
Add type to template function (#82564)

bieniu committed:
Translate `PowerOff` state of `select` entity in Harmony integration (#77491)

* Add custom device_class for select entity

* Update tests

* Make the state PowerOff translatable

* Update strings.select.json file

* add select.en.json

Co-authored-by: J. Nick Koston <nick@koston.org>

GitHub-Action committed:
[ci skip] Translation update

mkmer committed:
Add reauth config flow to Whirlpool (#82532)

* Add ReauthFlow

* Update homeassistant/components/whirlpool/config_flow.py

Darn it - thought I caught all of these.

Co-authored-by: Abílio Costa <abmantis@users.noreply.github.com>

* Update homeassistant/components/whirlpool/config_flow.py

Co-authored-by: Abílio Costa <abmantis@users.noreply.github.com>

* Update homeassistant/components/whirlpool/config_flow.py

Co-authored-by: Abílio Costa <abmantis@users.noreply.github.com>

* Update homeassistant/components/whirlpool/config_flow.py

Co-authored-by: Abílio Costa <abmantis@users.noreply.github.com>

* Update homeassistant/components/whirlpool/config_flow.py

Co-authored-by: Abílio Costa <abmantis@users.noreply.github.com>

* Update homeassistant/components/whirlpool/config_flow.py

Co-authored-by: Abílio Costa <abmantis@users.noreply.github.com>

Co-authored-by: Abílio Costa <abmantis@users.noreply.github.com>

ludeeus committed:
Do not delete issue when aborting repairs fix flow (#82593)

* Do not delete issue when aborting repairs fix flow

* lint

* types

epenet committed:
Migrate precipitation units to an enum (#81143)

jbouwh committed:
Strict type hints for MQTT integration (#82317)

* Strict type hints for MQTT integration

* Fix errors

* Additional corrections

* Use cv.template to avoid untyped calls

* Enable strict typing policy for MQTT integration

* Use ignore[no-untyped-call]

* Use # type: ignore[unreachable]

* Correct cast

* Refactor getting discovery_payload

* Remove unused type ignore comments

OnFreund committed:
Allow `device_attr` and `is_device_attr` to be used as a filter and a test (respectively) (#81924)

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

epenet committed:
Add type hints to template result wrapper (#82575)

* Add type hints to template result wrapper

* Remove `ignore[call-arg]`

* Use tuple

* Alphabetise

frenck committed:
Merge branch 'master' into dev

epenet committed:
Add OptionsFlow helper class (#82531)

* Add OptionsFlow helper classes

* More integrations

* Adjust SchemaOptionsFlowHandler

* Use single class

* Simplify access to options

* Reduce PR

* Make _options private

* Add test

cdce8p committed:
Fix round typing [homewizard] (#82628)

epenet committed:
Set last_step in SchemaCommonFlowHandler (#82616)

* Set last_step in SchemaCommonFlowHandler

* Always use boolean

* Adjust next_step definition

cdce8p committed:
Fix round typing [mqtt] (#82629)

epenet committed:
Make async_options_flow_finished optional (#82615)

* Make async_options_flow_finished optional

* Adjust docstring

ludeeus committed:
Add repair abort flow to demo integration (#82614)

epenet committed:
Add type hint to template wrapper (#82563)

* Add type to template hassfunction decorator

* Adjust to use EvalContext

* Use runtime.Context

* Use TypeVar for context

* Use jinja2.runtime.Context

* Reverse declarations

* Use Any

* Update homeassistant/helpers/template.py

Co-authored-by: Marc Mueller <30130371+cdce8p@users.noreply.github.com>

Co-authored-by: Marc Mueller <30130371+cdce8p@users.noreply.github.com>

rogelio-o committed:
Add unknown tuya modes as presets (#82056)

Fixes https://github.com/home-assistant/core/issues/81681
fixes undefined

allenporter committed:
Bump gcal_sync to 4.0.3 (#82606)

fixes undefined

emontnemery committed:
Fix show_menu support in SchemaFlowFormStep (#82634)

* Fix show_menu support in SchemaFlowFormStep

* Add test

* Fix test

frenck committed:
Remove Google Chat/Hangouts integration (#82645)

epenet committed:
Allow next_step to be a string (#82647)

* Allow next_step to be a string

* Adjust docstrings

* Add test

thecode committed:
Add Shelly tests coverage (#82642)

* Add Shelly tests coverage

* Review comments

* Remove leftovers

rappenze committed:
Support hvacsystem in fibaro integration (#78234)

fixes undefined

felipecrs committed:
Update base image to 2022.11.0 (#82527)

fixes undefined

bieniu committed:
Bump nextdns library to 1.2.0 (#82643)

epenet committed:
Improve Selector typing (#82636)

raman325 committed:
Add `text` platform (#79454)

Co-authored-by: Franck Nijhof <frenck@frenck.nl>
Co-authored-by: Franck Nijhof <git@frenck.dev>

thecode committed:
Mark Shelly as Platinum integration (#82656)

epenet committed:
Use SchemaOptionsFlowHandler in Accuweather (#82627)

Use SchemaOptionsFlowHandler in accuweather

gjohansson-ST committed:
Add myself as codeowner for min_max (#82658)

codeowner

gjohansson-ST committed:
Add sum to min_max helper (#82651)

sum to min_max

emontnemery committed:
Fix docstring in SchemaFlowFormStep (#82612)

thecode committed:
Fix Shelly gen2 channel name (#82655)

* Fix Shelly gen2 channel name

* Review comment

gjohansson-ST committed:
Code quality Local IP (#82050)

rytilahti committed:
Do not crash on non-existing enum values in xiaomi_miio.select (#82625)

fixes undefined

gjohansson-ST committed:
Add unique id for Scrape config entry entities (#82508)

* scrape unique id

* fix uuid str

* add back UoM

tkdrob committed:
Remove deprecated Radarr YAML config (#81210)

bdraco committed:
Fix un-retrieved future in esphome ble client when library raises (#82537)

zhulik committed:
Bump aiotractive, properly handle auth errors when reading events (#76715)

* Bump aiotractive, start handling authorzation errors when reading events

* Bump aiotractive, start handling authorzation errors when reading events

* Properly handle unauthorized errors

* Update homeassistant/components/tractive/__init__.py

Co-authored-by: J. Nick Koston <nick@koston.org>

* Use await when unsibscribing after auth error

* Update homeassistant/components/tractive/__init__.py

Co-authored-by: J. Nick Koston <nick@koston.org>

Co-authored-by: J. Nick Koston <nick@koston.org>

epenet committed:
Simplify SchemaFlowStep typing (#82661)

* Simplify SchemaFlowStep typing

* Adjust accuweather

emontnemery committed:
Fire events when long term statistics is updated (#82492)

* Fire events when long term statistics is updated

* Allow the new events to be subscribed to by anyone

* Address review comments

* Finish renaming events

* Finish renaming events

* Fix do_adhoc_statistics

* Adjust tests

* Adjust tests

emontnemery committed:
Pass sensor DOMAIN constant to EntitySelectorConfig (#82670)

gjohansson-ST committed:
Selector for SingleEntitySelectorConfig allow domain to be a list (#82666)

uvjustin committed:
Fix mime type in forked_daapd test (#82660)

emontnemery committed:
Add integration filter to utility_meter.calibrate service (#82671)

bdraco committed:
Fix iBeacons with infrequent random mac address changes unexpectedly going unavailable (#82668)

fixes https://github.com/home-assistant/core/issues/79781

emontnemery committed:
Allow configuring country and language in core config (#81734)

* Allow configuring country and language in core config

* Add script for updating list of countries

* Use black for formatting

* Fix quoting

* Move country codes to a separate file

* Address review comments

* Add generated/countries.py

* Get default language from owner account

* Remove unused variable

* Add script to generate list of supported languages

* Add tests

* Fix stale docsring

* Use format_python_namespace

* Correct async_user_store

* Improve typing

* Fix with_store decorator

* Initialize language in core store migration

* Fix startup

* Tweak

* Apply suggestions from code review

Co-authored-by: Franck Nijhof <git@frenck.dev>

* Update storage.py

Co-authored-by: Franck Nijhof <git@frenck.dev>

GitHub-Action committed:
[ci skip] Translation update

epenet committed:
Add handler to validate_user_input (#82681)

* Add handler to validate_user_input

* Adjust group config flow

jbouwh committed:
Announce a deprecated callback for MQTT subscribe will stop working (#82610)

* Announce a deprecated callback will stop working

* Use 2 months grace period

* Use 2 months grace period

* Use 2 months grace period

* Add deprecation comment to tests

epenet committed:
Simplify schema callback in SchemaFlowFormStep (#82682)

* Simplify SchemaFlowFormStep.schema callback

* Expose parent handler

* Adjust docstrings

bouwew committed:
Bump plugwise to v0.25.12 (#82146)

Co-authored-by: Franck Nijhof <frenck@frenck.nl>
fixes undefined

epenet committed:
Expose config_entry and options as properties (#82691)

Expose config_entry and options as a property

frenck committed:
Add support for raising ConfigEntryError (#82689)

eifinger committed:
Use SensorDeviceClass.DISTANCE for here_travel_time (#79159)

* Use SensorDeviceClass.DISTANCE in here_travel_time

* Removed unused import

* Use explicit units and conversion

gjohansson-ST committed:
Add number entities to config flow for min_max (#82665)

jbouwh committed:
Set mqtt quality scale to gold (#82696)

epenet committed:
Fix None schema in SchemaCommonFlowHandler (#82699)

epenet committed:
Rename options to suggested_values (#82700)

klaasnicolaas committed:
Add API key validation for Forecast.Solar (#80856)

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>
Co-authored-by: Franck Nijhof <git@frenck.dev>
Co-authored-by: Franck Nijhof <frenck@frenck.nl>

epenet committed:
Merge test files (#82703)

frenck committed:
Centralize exception handling in Plugwise (#82694)

emontnemery committed:
Allow SchemaFlowFormStep.next_step to return None (#82707)

epenet committed:
Use SchemaOptionsFlowHandler in apple_tv (#82688)

epenet committed:
Use SchemaOptionsFlowHandler in aurora (#82687)

epenet committed:
Use SchemaOptionsFlowHandler in airvisual (#82686)

jbouwh committed:
Clean up or correct from removed deprecation for MQTT platform (#82684)

Clean up or correct from removed deprecation

OnFreund committed:
Add exception trace in Risco config flow (#82683)

uvjustin committed:
Fix owntone websocket reconnection (#82680)

epenet committed:
Use SchemaOptionsFlowHandler in aemet (#82638)

* Use SchemaOptionsFlowHandler in aemet

* Update homeassistant/components/aemet/config_flow.py

* Update config_flow.py

* black

epenet committed:
Add callback to SchemaFlowFormStep for suggested_values (#82706)

* Allow callback for suggested_values

* docstring

epenet committed:
Use SchemaOptionsFlowHandler in azure-event-hub (#82619)

* Use SchemaOptionsFlowHandler in azure-event-hub

* Update config_flow.py

* black

emontnemery committed:
Fix docstring for entity helper (#82704)

Fix some language in entity.py

emontnemery committed:
Tweak scaffold script (#82705)

jbouwh committed:
Correction in mqtt error string (#82712)

Text correction in mqtt error string

bdraco committed:
Fix grammar in ibeacon comments (#82718)

* Fix grammar in ibeacon comments

* Update homeassistant/components/ibeacon/coordinator.py

GitHub-Action committed:
[ci skip] Translation update

liudger committed:
Add diagnostics to bsblan (#80957)

* add diagnostics to bsblan

* add device info

check firmware.

* add test for diagnostics

* Update tests/components/bsblan/test_diagnostics.py

Co-authored-by: Aarni Koskela <akx@iki.fi>

* add fixture for diagnostics test

Co-authored-by: Aarni Koskela <akx@iki.fi>

mbo18 committed:
Add more sensors to SensorEntityDescription for RFLink (#82036)

* Add more sensors to SensorEntityDescription

* changes from comments

* add device_class precipitation

* fix test

* change state_class for total_rain

farmio committed:
Update xknx to 2.0.0 (#82709)

Kane610 committed:
Fix deCONZ Air Quality PPB device class and unit of measurement (#82726)

Fix Air Quality PPB device class and unit of measurement

janiversen committed:
Solve modbus binary slave problem (#82338)

* Solve modbus binary slave problem.

OnFreund committed:
Reduce error trace in Risco config flow to debug (#82730)

mbo18 committed:
Add native unit where device class is set in rflink (#82729)

Add native unit where device class is set

emontnemery committed:
Add helper to calculate statistic period start and end (#82493)

* Add helper to calculate statistic period start and end

* Don't parse values in resolve_period

* Add specific test for resolve_period

* Improve typing

* Move to recorder/util.py

* Extract period schema

gjohansson-ST committed:
Fix pushbullet test that break CI (#82744)

pushbullet break ci

bieniu committed:
Bump `nextdns` library to version 1.2.2 (#82735)

Bump nextdns library to version 1.2.2

AlePerla committed:
Integration of DP Instructions support for WK devices if set during config
Pylint warnings fix

AlePerla committed:
Merge pull request #1 from home-assistant/dev

merging main repository changes inside dev

AlePerla committed:
Merge branch 'dev' into tuya_thermostat_eco-mode

AlePerla committed:
Constants for Instruction types strings
Restored mode since it was not working
